### ISSUE TITLE
Customizable statusline

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#708]
Following some discussion on #649, it might be a good idea to allow for a configurable statusline through templating. We can borrow some code from [here](https://github.com/mitsuhiko/indicatif/blob/27c8cc3078b65aa2573dbe673a00471d6662698c/src/style.rs#L238-L390), and make some of the changes suggested by Blaz: 'We might be able to reuse that with some tweaking (lazy_static -> once_cell, disabling styles for now).'

**Comment by Omnikar** [comment]
What should be used in the template string to indicate the separation between the left and right sides of the statusline? Maybe `{|}`?

**Comment by kirawi** [comment]
Maybe we could use the alignment operators: `<`, `>`? It would just be relative to the left and right sides of the status line.

**Comment by Omnikar** [comment]
If we just use a "separator", such as `{|}`, like I mentioned, we can just allow `left{|}center{|}right`.

**Comment by kirawi** [comment]
Sounds good.

**Comment by Omnikar** [comment]
Though we'd have to decide what to do if there are less than 2 separators present in the format string. I think:
- 0: `left`
- 1: `left{|}right`
If you wanted only left and center or right and center, you could do `left{|}center{|}` or `{|}center{|}right`. And center-only can be achieved via `{|}center{|}`.

**Comment by gbaranski** [comment]
Or we could do
```toml
[statusline]
left = "editing {filename}"
center = "Date: {date}"
right = "LSP: {lsp-status}"
```

What do you think about that?

**Comment by kirawi** [comment]
Conditional templating would also be interesting, e.g. #447

**Comment by gbaranski** [comment]
I3status-rust does something similar to what we want. https://github.com/greshake/i3status-rust/blob/master/doc/blocks.md

**Comment by sudormrfbin** [comment]
> Maybe something like the C++ short hand if else statement (`(condition) ? expressionTrue : expressionFalse`) with some predefined conditions?

I think we want to avoid introducing constructs like that since it will probably spiral into injecting more and more logic handling into a plain toml file and before we know it we would have invented a mini language that's similar to most things out there but also very different from most things out there. It was also brought up in Matrix a while ago and there was a (more or less) general consensus that it's better left to when we have a proper plugin system in place which can be used to dynamically define the config file or an actual configuration _programming_ language (lisps were of interest last time IIRC).

**Comment by bestouff** [comment]
I'd like to have this kind of eye candy in the status bar possible:

![image](https://user-images.githubusercontent.com/1163188/143600098-852c1cdd-0ff8-41df-96c2-db7db744758a.png)
Here we see the git branch, file name, modification status, git lines added/changed/removed, LSP errors/warnings, in-tree indicator, LSP server and lang, and "scrollbar" position.  
As showed here it uses nerd fonts, so that probably requires a config item to degrade from nerd font to plain unicode.

**Comment by NNBnh** [comment]
This look like a status line from Emacs, looking good! I have composed a list of component for the statusline taking from [Kakoune](https://github.com/mawww/kakoune/wiki/Status-line), [Vim's airline](https://github.com/vim-airline/vim-airline) and that Emacs's line above:
- [ ] Current position information:
  - [X] Cursor line.
  - [X] Cursor column.
  - [X] Cursor display column.
  - [ ] [Cursor position as a percentage of the whole file](https://github.com/mawww/kakoune/wiki/Status-line#cursor-position-as-a-percentage-of-the-whole-file) (a more simple approach to the "scrollbar position" that show on the Emacs screenshot).
  - [ ] Cursor count.
- [X] Current mode.
- [ ] Buffer's line count.
- [X] Filename.
- [ ] Filepath.
- [ ] Read-only flag (Helix currently doesn't support read-only buffer).
- [X] Modification status.
- [ ] File encoding (e.g: "utf-8").
- [ ] Fileformat (e.g: "unix").
- [ ] Filetype (LSP server and lang).
- [X] LSP errors/warnings.
- [ ] VCS information (git added/changed/removed).
- [ ] In-tree indicator (idk what this is...)

**Comment by bestouff** [comment]
This is the status line from Lunarvim (an easy to install full-fledged nvim config) but yeah.

**Comment by pickfire** [comment]
One more which I find useful is the search position from doom emacs, `10/20` like that.

**Comment by hahanein** [comment]
IntelliJ has a very simple status line configuration feature: https://www.jetbrains.com/webstorm/guide/tips/customize-status-bar/

It'd look a little like this:

```
// config.toml
[editor]
status-line = ["git-branch", "file-encoding", "search-position"]
```

VSCode is similarly restrictive. Furthermore, _a plugin author_ may add status bar items [using this API](https://code.visualstudio.com/api/references/vscode-api#StatusBarItem) (alignment configurable).

In my mind this is really user friendly. Adding and removing features is straight-forward. No knowledge of a template language is required. No need to debug layouts (maybe _less_ need to debug layouts).

If you want more control; if a plugin author wants to provide you with more control, they should be able to take full control over status line, though. However, in that case I'm not sure you'd want to impose restrictions such as the use of a specific template language or of specific alignment primitives (left, center, right).

A couple more data points:

- Kakoune appears to restrict customization to the right side of the status line: https://github.com/mawww/kakoune/wiki/Status-line
- As a user, adding git support to Vim's status line is rather painful in my mind: https://shapeshed.com/vim-statuslines

**Comment by Zoybean** [comment]
One thing I liked from my vim config was displaying the hex value of the character under the (primary, in helix' case) cursor. I'd love to have that in helix

**Comment by kirawi** [comment]
Resolved by #2434

**Comment by Rudxain** [comment]
> displaying the hex value of the character under the (primary, in helix' case) cursor

```toml
[keys.normal]
"C-h" = ":! printf %%s '%{selection}' | od -A n -t x1"
[keys.select]
"C-h" = ":echo %sh{printf %%s '%{selection}' | od -A n -t x1}"
```
it [should work on any POSIX system](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/od.html) (no dep on `hd` or `xxd`). But it won't work if `selection` contains `'`

(to everyone) I'm sorry for necro-posting
